File: native/android/MotherAIResourceGuardian.kt
package ai.laneriq.runtime

import android.content.Context
import android.net.ConnectivityManager
import android.os.BatteryManager
import android.os.Build
import android.os.PowerManager

/**
 * Native telemetry adapter for a future signed LANERIQ Android shell.
 *
 * This class does not create long-running background services and does not bypass Android
 * power management. The host app must use system-managed work (normally WorkManager) and
 * pass the lease/constraint state into snapshot(). Mobile Community Compute stays disabled.
 */
class MotherAIResourceGuardian(private val context: Context) {
    companion object {
        const val BRIDGE_VERSION = "2026-09-05.1"
    }

    fun snapshot(
        lifecycleState: String,
        backgroundLease: String = "none",
        backgroundConstraintsSatisfied: Boolean = false,
        userActive: Boolean = true,
        deviceIdle: Boolean = false,
        networkType: String = "unknown"
    ): Map<String, Any?> {
        val power = context.getSystemService(Context.POWER_SERVICE) as PowerManager
        val battery = context.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
        val connectivity = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

        val batteryPercent = battery.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
        val chargingStatus = battery.getIntProperty(BatteryManager.BATTERY_PROPERTY_STATUS)
        val charging = chargingStatus == BatteryManager.BATTERY_STATUS_CHARGING || chargingStatus == BatteryManager.BATTERY_STATUS_FULL

        return mapOf(
            "bridgeVersion" to BRIDGE_VERSION,
            "platform" to "android",
            "thermalState" to thermalState(power),
            "lowPowerMode" to power.isPowerSaveMode,
            "batteryLevel" to if (batteryPercent in 0..100) batteryPercent / 100.0 else null,
            "charging" to charging,
            "lifecycleState" to lifecycleState,
            "backgroundLease" to backgroundLease,
            "backgroundConstraintsSatisfied" to backgroundConstraintsSatisfied,
            "networkType" to networkType,
            "meteredNetwork" to connectivity.isActiveNetworkMetered,
            "userActive" to userActive,
            "deviceIdle" to deviceIdle,
            "communityComputeAllowed" to false,
            "downloadedExecutableWorkloadsAllowed" to false
        )
    }

    private fun thermalState(power: PowerManager): String {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) return "unknown"
        return when (power.currentThermalStatus) {
            PowerManager.THERMAL_STATUS_NONE, PowerManager.THERMAL_STATUS_LIGHT -> "nominal"
            PowerManager.THERMAL_STATUS_MODERATE -> "fair"
            PowerManager.THERMAL_STATUS_SEVERE -> "serious"
            PowerManager.THERMAL_STATUS_CRITICAL,
            PowerManager.THERMAL_STATUS_EMERGENCY,
            PowerManager.THERMAL_STATUS_SHUTDOWN -> "critical"
            else -> "unknown"
        }
    }
}

File: native/ios/MotherAIResourceGuardian.swift
import Foundation
import UIKit

/// Native telemetry adapter for a future signed LANERIQ iOS/iPadOS shell.
///
/// This module intentionally does not schedule background work by itself. The host app must
/// obtain legitimate system-managed background runtime (for example BGProcessingTask or
/// task-completion time) and then pass that lease state to `snapshot`.
///
/// No Community Compute execution is permitted on iOS/iPadOS.
public final class MotherAIResourceGuardian {
    public static let bridgeVersion = "2026-09-05.1"

    public init() {
        UIDevice.current.isBatteryMonitoringEnabled = true
    }

    public func snapshot(
        lifecycleState: String,
        backgroundLease: String = "none",
        backgroundConstraintsSatisfied: Bool = false,
        userActive: Bool = true,
        deviceIdle: Bool = false,
        networkType: String = "unknown",
        meteredNetwork: Bool = false
    ) -> [String: Any] {
        let process = ProcessInfo.processInfo
        let device = UIDevice.current

        return [
            "bridgeVersion": Self.bridgeVersion,
            "platform": UIDevice.current.userInterfaceIdiom == .pad ? "ipados" : "ios",
            "thermalState": thermalState(process.thermalState),
            "lowPowerMode": process.isLowPowerModeEnabled,
            "batteryLevel": normalizedBatteryLevel(device.batteryLevel),
            "charging": device.batteryState == .charging || device.batteryState == .full,
            "lifecycleState": lifecycleState,
            "backgroundLease": backgroundLease,
            "backgroundConstraintsSatisfied": backgroundConstraintsSatisfied,
            "networkType": networkType,
            "meteredNetwork": meteredNetwork,
            "userActive": userActive,
            "deviceIdle": deviceIdle,
            "communityComputeAllowed": false,
            "downloadedExecutableWorkloadsAllowed": false
        ]
    }

    private func normalizedBatteryLevel(_ level: Float) -> Any {
        guard level >= 0 else { return NSNull() }
        return max(0, min(1, Double(level)))
    }

    private func thermalState(_ state: ProcessInfo.ThermalState) -> String {
        switch state {
        case .nominal:
            return "nominal"
        case .fair:
            return "fair"
        case .serious:
            return "serious"
        case .critical:
            return "critical"
        @unknown default:
            return "unknown"
        }
    }
}
